README.md
DB optimized for a bunch of PNG images.

Idea
====

The idea is to split PNG images into many blocks and have each block stored in a DB.
If there are several equal blocks, it is only stored once.
Via a hash table, the lookup for such blocks is made fast.

Use case
========

I am collecting screenshots (for several reasons; one is to play around with machine learning / computer vision;
one example is here: <https://github.com/albertz/screenshooting>). A lot of them. :)

Right now, I have about 88k screenshots with about 77GB. And as many of them have a lot of repetitive areas
(on some days, I were making a screenshot every 10 seconds, even when not using the computer at all,
so the only changing part was the time display), I didn't wanted to waste so much space on so much repetitive data.

With this PNG DB, I have a compression rate of about 400-500%
(for the first 1k screenshots or so; probably the rate will even be higher for all of them).

This example with the screenshots is probably an extreme case (where this applies extremely well).
But I guess in many other cases where you are collecting a huge amount of PNG images
(with computer-generated content; real-world images would not work that well), you can safe some space by it.

And if this gets optimized as far as possible, it may be even faster than normal filesystem access
(because of less disk IO).

Technical details
=================

To make things easier on the PNG side, it just parses down until it gets a scanline serialization.
Multiple directly following scanline (of same width) serializations parts build up a block
(so it actually really matches a block in the real picture). But I don't do any of the PNG filtering.
PNG spec: <http://www.w3.org/TR/PNG/>

The general DB layout is as follows:

- ("data." unique id -> zlib compressed data) data pairs
- ("sha1refs." SHA1 -> set of ids) data pairs
- ("fs." filename -> id) data pairs

Such data value (uncompressed) starts with a data-type-byte. Only 3 types are there currently:

- PNG file summary
- PNG chunk (all non-data PNG chunks)
- PNG block

There are multiple DB backend implementations:

- The filesystem itself. But creates a lot of files!
- [Redis](http://redis.io/). Via [hiredis](https://github.com/antirez/hiredis). As everything is in memory, you are a bit limited.
- [KyotoCabinet](http://fallabs.com/kyotocabinet/). (Currently the default.)

Comparison with other compression methods / deduplicators
=========================================================

In the beginning, I thought about using some generic image library to be able to handle just any
image type and then operate just on the raw data. This would even give me some slight better compression rate
because now, I am operating on PNGs scanline serializations and there are 5 different ways (filters) in PNG
to represent a scanline.

However, because I am storing all the data as PNG raw data in the DB, the reconstruction of the PNG should
be much faster. In the more generic case, I would have to recompress/reencode the PNG. Now I only have
to (roughly) collect and glew the parts together and run the PNG zlib over it.

Using a general deduplicator / compressor on the raw data (uncompressed PNG, TGA or BMP):
It would be based on connected chunks of data; i.e., in the image, it would mean one or many following scanlines.
But what I am doing is based on rectangular blocks in the image. So I am able to get much bigger
chunks of data which is repetitive.

Something like what is done in video compressions methods like H264: This might actually be a very good idea.
And it should be possible to just add it now to my current method.

Tools
=====

It comes with several tools. Some of them:

- db-push: Pushes a single PNG into the DB.
- db-push-dir: Pushes all PNGs in a given directory into the DB.
- db-extract-file: Extracts a single PNG from the DB.
- db-fuse: Simple FUSE interface to the DB. (Slow though because it is not very optimized!)

Compilation
===========

Just run `./compile.sh`.

For Mac: If you haven't MacFUSE installed, install it from here: <http://code.google.com/p/macfuse/>

TODOs
=====

- Many parts could be optimized a lot.
- Try with other DB backend implementations.
Maybe [mongoDB](http://www.mongodb.org/) or [Basho Riak](http://www.basho.com/products_riak_overview.php).
Or improve the filesystem implementation (which is incomplete anyway currently).
- To make the FUSE interface faster, the caching must be improved.
Also, there should be a way to get the filesize in advance and maybe also to seek in constant time.
Probably, to make this possible, we need to have a fixed compression algorithm and
the file summary must contain some offset information.
- We could also store other image formats in a similar way. And also general files.
There should be also a standard fallback.
- The FUSE interface could also support writing.


-Albert Zeyer, <http://www.az2000.de>


kyotocabinet/ChangeLog
2011-03-01  FAL Labs  <info@fallabs.com>

	* kclangc.h (kcmapnew, kcmapdel, kcmapiterator, kcmapsorter): new functions.

	* kclangctest.cc (runmap, procmap): new functions.

	- Release: 1.2.48

2011-02-25  FAL Labs  <info@fallabs.com>

	* kcdb.h (BasicDB::occupy): new function.

	- Release: 1.2.47

2011-02-23  FAL Labs  <info@fallabs.com>

	* kcutil.h (SIZEMAX, FLTMAX, DBLMAX): new constants.

	* kcplantdb.h (PlantDB::open): repair mechanism was modified.

	* kcplantdb.h (PlantDB::Cursor::accept_spec): a bug related to transactin was fixed.

	- all database classes were refactored to support some old compilers.

	- Release: 1.2.46

2011-02-21  FAL Labs  <info@fallabs.com>

	* kcdbext.h (MapReduce::execute): adaptive comparator was added.

	* kcdbext.h (MapReduce::execute_reduce): new function.

	* kcdbext.h (MapReduce::before, MapReduce::after): new functions.

	- Release: 1.2.45

2011-02-15  FAL Labs  <info@fallabs.com>

	* kcdb.h (DB::Visitor::visit_before, DB::Visitor::visit_after): new functions.

	- configuration files were modified for affinity for binary distributions.

	- all classes were refactored to abolish unnamed namespaces.

	- Release: 1.2.44

2011-02-12  FAL Labs  <info@fallabs.com>

	* kchashdb.h (HashDB): the type of record locking objects was changed.

	* kcdirdb.h (DirDB): the type of record locking objects was changed.

	* kcplantdb.h (PlantDB): the type of page locking objects was changed.

	- Release: 1.2.43

2011-02-10  FAL Labs  <info@fallabs.com>

	* kcdb.h (DB::Cursor::get_key, DB::Cursor::get_value): signature was modified.

	* kcdb.h (DB::Cursor::get_pair, DB::Cursor::get): renamed.

	* kcdb.h (DB::get): signature was modified.

	- Release: 1.2.42

2011-02-07  FAL Labs  <info@fallabs.com>

	* kccommon.h: macros for integer range were added.

	* kchashdb.h (HashDB::begin_transaction): behavior of locking was modified.

	* kcdirdb.h (DirDB::begin_transaction): behavior of locking was modified.

	- Release: 1.2.41

2011-02-05  FAL Labs  <info@fallabs.com>

	* kccompare.h (LexicalDescendingComparator, LexicalDescendingComparator): new classes.

	* kccompare.h (LEXICALCOMP, DECIMALCOMP): the type was changed to pointer.

	* kccompress.h (ZLIBRAWCOMP): the type was changed to pointer.

	* kcplantdb.h (PlantDB::load_meta): a bug related to custom comparator was fixed.

	- Release: 1.2.40

2011-02-01  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::open): checking type type was added.

	* kcplantdb.h (Plantdb::reorganize_file): recovery condition was modified.

	* kcpolymgr.cc (proccheck): a bug related to path handling was fixed.

	- Release: 1.2.39

2011-01-30  FAL Labs  <info@fallabs.com>

	* kcthread.cc: threshold of sleep locking was modified.

	* kcplantdb.h (PlantDB::accept): a bug related to race condition was fixed.

	* kcplantdb.h (PlantDB::save_leaf_node): locking mode was modified.

	- Release: 1.2.38

2011-01-25  FAL Labs  <info@fallabs.com>

	* kcutil.h (strvecdump, strvecload, strmapdump, strmapload): new functions.

	- Release: 1.2.37

2011-01-21  FAL Labs  <info@fallabs.com>

	* kcutil.cc (getsysinfo): a bug related to memory stats was fixed.

	* kcthread.cc (Thread::chill): new function.

	* kcthread.cc (ScopedMutex, ScopedSpinLock): refactored using primitives.

	* kcthread.cc (ScopedRWLock, ScopedSpinRWLock): refactored using primitives.

	- Release: 1.2.36

2011-01-11  FAL Labs  <info@fallabs.com>

	* kcutil.cc: the order of header inclusion was modified.

	* kcthread.cc (SpinRWLock): space efficiency was improved.

	- Release: 1.2.35

2011-01-09  FAL Labs  <info@fallabs.com>

	* kcplantdb.h (PlantDB::accept, PlantDB::accept_bulk): behavior of async was modified.

	- Release: 1.2.34

2011-01-05  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::open): handling the WAL file was modified.

	* kchashdb.h (HashDB::clear): handling the open flag was added.

	* kcplantdb.h (PlantDB::clear): handling the recovery flag was modified.

	- Release: 1.2.33

2011-01-03  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::read_fast): error check was added.

	* kclangc.cc (kcdbacceptbulk, kcdbsetbulk, kcdbremovebulk, kcdbgetbulk): new functions.

	- Release: 1.2.32

2010-12-30  FAL Labs  <info@fallabs.com>

	* kcdb.h (BasicDB::accept_bulk): new function.

	* kcdb.h (BasicDB::set_bulk, BasicDB::remove_bulk, BasicDB::get_bulk): new functions.

	* kcthread.h (SlottedMutex::unlock_all): the order became forward.

	* kcpolymgr.cc (procsetbulk, procremovebulk, procgetbulk): new functions.

	- Release: 1.2.31

2010-12-19  FAL Labs  <info@fallabs.com>

	* kcpolydb.h (PolyDB::set_internal_db): new function.

	- Release: 1.2.30

2010-12-11  FAL Labs  <info@fallabs.com>

	* kccommon.h (modfl): new function for portability to cygwin.

	* kcthread.cc (Mutex::lock_try): cygwin was added to work around cases.

	* cmdcommon.h (oprintf): new function instead of "iprintf".

	- Release: 1.2.29

2010-12-10  FAL Labs  <info@fallabs.com>

	* ktutil.h (atoin, atofn): new functions.

	- Release: 1.2.28

2010-11-30  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::write, File::read): warnings on 32-bit environment were cleared.

	* kcfile.cc (File::write_file, File::make_directory): modified for Win32 virus checkers.

	- Release: 1.2.27

2010-11-28  FAL Labs  <info@fallabs.com>

	* ktutil.h (vstrprintf): solved a compilation warning.

	* kcdbext.h (MapReduce::~MapReduce): modified as a virtual function.

	- Release: 1.2.26

2010-11-17  FAL Labs  <info@fallabs.com>

	* kcdb.h (BasicDB::MetaTrigger): new class.

	* kcdb.h (BasicDB::tune_trigger): new function.

	* kcpolydb.h (PolyDB::StreamMetaTrigger): new class.

	- Release: 1.2.25

2010-11-10  FAL Labs  <info@fallabs.com>

	* kcstashdb.h, kcstashdb.o: new files.

	- Release: 1.2.24

2010-11-06  FAL Labs  <info@fallabs.com>

	* kcutil.h (sizevarnum): new function.

	* kcmap.h (TinyHashMap): new class.

	* kcdbext.h (MapReduce::execute): the cache algorithm was modified.

	- Release: 1.2.23

2010-11-04  FAL Labs  <info@fallabs.com>


Sha1.cpp
/*
SHA-1 in C
By Steve Reid <sreid@sea-to-sky.net>
100% Public Domain

-----------------
Modified 7/98
By James H. Brown <jbrown@burgoyne.com>
Still 100% Public Domain

Corrected a problem which generated improper hash values on 16 bit machines
Routine SHA1Update changed from
	void SHA1Update(SHA1_CTX* context, unsigned char* data, unsigned int
len)
to
	void SHA1Update(SHA1_CTX* context, unsigned char* data, unsigned
long len)

The 'len' parameter was declared an int which works fine on 32 bit machines.
However, on 16 bit machines an int is too small for the shifts being done
against
it.  This caused the hash function to generate incorrect values if len was
greater than 8191 (8K - 1) due to the 'len << 3' on line 3 of SHA1Update().

Since the file IO in main() reads 16K at a time, any file 8K or larger would
be guaranteed to generate the wrong hash (e.g. Test Vector #3, a million
"a"s).

I also changed the declaration of variables i & j in SHA1Update to
unsigned long from unsigned int for the same reason.

These changes should make no difference to any 32 bit implementations since
an
int and a long are the same size in those environments.

--
I also corrected a few compiler warnings generated by Borland C.
1. Added #include <process.h> for exit() prototype
2. Removed unused variable 'j' in SHA1Final
3. Changed exit(0) to return(0) at end of main.

ALL changes I made can be located by searching for comments containing 'JHB'
-----------------
Modified 8/98
By Steve Reid <sreid@sea-to-sky.net>
Still 100% public domain

1- Removed #include <process.h> and used return() instead of exit()
2- Fixed overwriting of finalcount in SHA1Final() (discovered by Chris Hall)
3- Changed email address from steve@edmweb.com to sreid@sea-to-sky.net

-----------------
Modified 4/01
By Saul Kravitz <Saul.Kravitz@celera.com>
Still 100% PD
Modified to run on Compaq Alpha hardware.

-----------------
Modified 07/2002
By Ralph Giles <giles@ghostscript.com>
Still 100% public domain
modified for use with stdint types, autoconf
code cleanup, removed attribution comments
switched SHA1Final() argument order for consistency
use SHA1_ prefix for public api
move public api to sha1.h
*/

/*
Test Vectors (from FIPS PUB 180-1)
"abc"
  A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
  84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
A million repetitions of "a"
  34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
*/

#define SHA1HANDSOFF

#include <cstdio>
#include <cstring>

#include "Sha1.h"

static void SHA1_Transform(uint32_t state[5], const uint8_t buffer[64]);

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
/* FIXME: can we do this in an endian-proof way? */
#ifdef WORDS_BIGENDIAN
#define blk0(i) block->l[i]
#else
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&0xFF00FF00) \
    |(rol(block->l[i],8)&0x00FF00FF))
#endif
#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
    ^block->l[(i+2)&15]^block->l[i&15],1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);


#ifdef VERBOSE  /* SAK */
void SHAPrintContext(Sha1Context *context, char *msg){
  printf("%s (%d,%d) %x %x %x %x %x\n",
	 msg,
	 context->count[0], context->count[1],
	 context->state[0],
	 context->state[1],
	 context->state[2],
	 context->state[3],
	 context->state[4]);
}
#endif /* VERBOSE */

/* Hash a single 512-bit block. This is the core of the algorithm. */
static void SHA1_Transform(uint32_t state[5], const uint8_t buffer[64])
{
    uint32_t a, b, c, d, e;
    typedef union {
        uint8_t c[64];
        uint32_t l[16];
    } CHAR64LONG16;
    CHAR64LONG16* block;

#ifdef SHA1HANDSOFF
    static uint8_t workspace[64];
    block = (CHAR64LONG16*)workspace;
    memcpy(block, buffer, 64);
#else
    block = (CHAR64LONG16*)buffer;
#endif

    /* Copy context->state[] to working vars */
    a = state[0];
    b = state[1];
    c = state[2];
    d = state[3];
    e = state[4];

    /* 4 rounds of 20 operations each. Loop unrolled. */
    R0(a,b,c,d,e, 0); R0(e,a,b,c,d, 1); R0(d,e,a,b,c, 2); R0(c,d,e,a,b, 3);
    R0(b,c,d,e,a, 4); R0(a,b,c,d,e, 5); R0(e,a,b,c,d, 6); R0(d,e,a,b,c, 7);
    R0(c,d,e,a,b, 8); R0(b,c,d,e,a, 9); R0(a,b,c,d,e,10); R0(e,a,b,c,d,11);
    R0(d,e,a,b,c,12); R0(c,d,e,a,b,13); R0(b,c,d,e,a,14); R0(a,b,c,d,e,15);
    R1(e,a,b,c,d,16); R1(d,e,a,b,c,17); R1(c,d,e,a,b,18); R1(b,c,d,e,a,19);
    R2(a,b,c,d,e,20); R2(e,a,b,c,d,21); R2(d,e,a,b,c,22); R2(c,d,e,a,b,23);
    R2(b,c,d,e,a,24); R2(a,b,c,d,e,25); R2(e,a,b,c,d,26); R2(d,e,a,b,c,27);
    R2(c,d,e,a,b,28); R2(b,c,d,e,a,29); R2(a,b,c,d,e,30); R2(e,a,b,c,d,31);
    R2(d,e,a,b,c,32); R2(c,d,e,a,b,33); R2(b,c,d,e,a,34); R2(a,b,c,d,e,35);
    R2(e,a,b,c,d,36); R2(d,e,a,b,c,37); R2(c,d,e,a,b,38); R2(b,c,d,e,a,39);
    R3(a,b,c,d,e,40); R3(e,a,b,c,d,41); R3(d,e,a,b,c,42); R3(c,d,e,a,b,43);
    R3(b,c,d,e,a,44); R3(a,b,c,d,e,45); R3(e,a,b,c,d,46); R3(d,e,a,b,c,47);
    R3(c,d,e,a,b,48); R3(b,c,d,e,a,49); R3(a,b,c,d,e,50); R3(e,a,b,c,d,51);
    R3(d,e,a,b,c,52); R3(c,d,e,a,b,53); R3(b,c,d,e,a,54); R3(a,b,c,d,e,55);
    R3(e,a,b,c,d,56); R3(d,e,a,b,c,57); R3(c,d,e,a,b,58); R3(b,c,d,e,a,59);
    R4(a,b,c,d,e,60); R4(e,a,b,c,d,61); R4(d,e,a,b,c,62); R4(c,d,e,a,b,63);
    R4(b,c,d,e,a,64); R4(a,b,c,d,e,65); R4(e,a,b,c,d,66); R4(d,e,a,b,c,67);
    R4(c,d,e,a,b,68); R4(b,c,d,e,a,69); R4(a,b,c,d,e,70); R4(e,a,b,c,d,71);
    R4(d,e,a,b,c,72); R4(c,d,e,a,b,73); R4(b,c,d,e,a,74); R4(a,b,c,d,e,75);
    R4(e,a,b,c,d,76); R4(d,e,a,b,c,77); R4(c,d,e,a,b,78); R4(b,c,d,e,a,79);

    /* Add the working vars back into context.state[] */
    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;
    state[4] += e;

    /* Wipe variables */
    a = b = c = d = e = 0;
}


/* SHA1Init - Initialize new context */
Sha1Context::Sha1Context()
{
    /* SHA1 initialization constants */
    state[0] = 0x67452301;
    state[1] = 0xEFCDAB89;
    state[2] = 0x98BADCFE;
    state[3] = 0x10325476;
    state[4] = 0xC3D2E1F0;
    count[0] = count[1] = 0;
}


/* Run your data through this. */
void Sha1Context::update(const char* d, size_t len)
{
	Sha1Context* context = this;
	const uint8_t* data = (uint8_t*)d;
    size_t i, j;

#ifdef VERBOSE
    SHAPrintContext(context, "before");
#endif

    j = (context->count[0] >> 3) & 63;
    if ((context->count[0] += len << 3) < (len << 3)) context->count[1]++;
    context->count[1] += (len >> 29);
    if ((j + len) > 63) {
        memcpy(&context->buffer[j], data, (i = 64-j));
        SHA1_Transform(context->state, context->buffer);
        for ( ; i + 63 < len; i += 64) {
            SHA1_Transform(context->state, data + i);
        }
        j = 0;
    }
    else i = 0;
    memcpy(&context->buffer[j], &data[i], len - i);

#ifdef VERBOSE
    SHAPrintContext(context, "after ");
#endif
}


/* Add padding and return the message digest. */
std::string Sha1Context::final()
{
	uint8_t digest[SHA1_DIGEST_SIZE];
	Sha1Context* context = this;
    uint32_t i;
    uint8_t  finalcount[8];

    for (i = 0; i < 8; i++) {
        finalcount[i] = (unsigned char)((context->count[(i >= 4 ? 0 : 1)]
         >> ((3-(i & 3)) * 8) ) & 255);  /* Endian independent */
    }
    update("\200", 1);
    while ((context->count[0] & 504) != 448) {
        update("\0", 1);
    }
    update((char*)finalcount, 8);  /* Should cause a SHA1_Transform() */
    for (i = 0; i < SHA1_DIGEST_SIZE; i++) {
        digest[i] = (uint8_t)
         ((context->state[i>>2] >> ((3-(i & 3)) * 8) ) & 255);
    }

    /* Wipe variables */
    i = 0;
    memset(context->buffer, 0, 64);
    memset(context->state, 0, 20);
    memset(context->count, 0, 8);
    memset(finalcount, 0, 8);	/* SWR */

#ifdef SHA1HANDSOFF  /* make SHA1Transform overwrite its own static vars */
    SHA1_Transform(context->state, context->buffer);
#endif
	
	return std::string((char*)digest, SHA1_DIGEST_SIZE);
}

/*
 // OpenSSL code
 
#include <openssl/sha.h>

std::string calc_sha1(const char* data, size_t size) {
	std::string s(SHA_DIGEST_LENGTH, 0);
	SHA1((const unsigned char*) data, size, (unsigned char*) &s[0]);
	return s;
}
*/

std::string calc_sha1(const char* data, size_t size) {
	Sha1Context c;
	c.update(data, size);
	return c.final();
}
